internal/pkg/agent/cmd/run.go

```diff
@@ -557,3 +559,40 @@ func setupMetrics(
 func isProcessStatsEnabled(cfg *monitoringCfg.MonitoringConfig) bool {
 	return cfg != nil && cfg.HTTP.Enabled
 }
+
+// handleUpgrade checks if agent is being run as part of an
+// ongoing upgrade operation, i.e. being re-exec'd and performs
+// any upgrade-specific work, if needed.
+func handleUpgrade() error {
+	upgradeMarker, err := upgrade.LoadMarker()
+	if err != nil {
+		return fmt.Errorf("unable to load upgrade marker to check if Agent is being upgraded: %w", err)
+	}
+
+	if upgradeMarker == nil {
+		// We're not being upgraded. Nothing more to do.
+		return nil
+	}
+
+	// In v8.8.0, we introduced a new installation marker file to indicate that
+	// an Agent was running as installed. When an installed Agent that's older
+	// than v8.8.0 is upgraded, this installation marker file is not present.
+	// So, in such cases, we need to create it manually post-upgrade.
+	// Otherwise, the upgrade will be unsuccessful (see
+	// https://github.com/elastic/elastic-agent/issues/2645).
+
+	// Only an installed Elastic Agent can be self-upgraded. So, if the
+	// installation marker file is already present, we're all set.
+	if info.RunningInstalled() {
+		return nil
+	}
+
+	// Otherwise, we're being upgraded from a version of an installed Agent
+	// that didn't use an installation marker file (that is, before v8.8.0).
+	// So create the file now.
+	if err := info.CreateInstallMarker(paths.Top()); err != nil {
```

Comment:
This has a bit of overlap with the upgrade migrations I am trying to implement in #2597 as I can see creating the install marker as part of an upgrade migration then the PrevVersion (yes I have to rely on the update marker to know which version we are coming from) is < 8.8.0

It's not something to do now but we could think of moving this part into its own migration

Reply:
Yes! While writing this I was thinking how it parallels DB migrations so, indeed, there's some overlap here.  ++ to incorporating this and #2597 into a more structured migrations process if that makes sense.